packages/taiko-client-rs/crates/test-harness/src/shasta/env.rs

```diff
@@ -0,0 +1,210 @@
+        event_indexer.wait_historical_indexing_finished().await;
+
+        // Build the proposer wired to the shared indexer and local codec.
+        let proposer_config = ProposerConfigs {
+            l1_provider_source: l1_source.clone(),
+            l2_provider_url: l2_http_url.clone(),
+            l2_auth_provider_url: l2_auth_url.clone(),
+            jwt_secret: jwt_secret_path.clone(),
+            inbox_address,
+            use_local_shasta_codec: true,
+            l2_suggested_fee_recipient,
+            propose_interval: Duration::from_secs(0),
+            l1_proposer_private_key,
+            gas_limit: None,
+        };
+        let proposer =
+            Arc::new(Proposer::new_with_indexer(proposer_config, event_indexer.clone()).await?);
+
+        info!(elapsed_ms = started.elapsed().as_millis(), "loaded ShastaEnv");
+        Ok(Self {
+            l1_source,
+            l2_http: l2_http_url,
+            l2_auth: l2_auth_url,
+            jwt_secret: jwt_secret_path,
+            inbox_address,
+            l2_suggested_fee_recipient,
+            l1_proposer_private_key,
+            taiko_anchor_address,
+            client_config,
+            client,
+            event_indexer,
+            proposer,
+            cleanup_provider,
+            snapshot_id,
+            blob_server: None,
+        })
+    }
+
+    /// Start a per-test blob server for serving the given sidecar.
+    pub async fn start_blob_server(&mut self, sidecar: BlobTransactionSidecar) -> Result<RpcUrl> {
+        let server = BlobServer::start(sidecar).await?;
+        let endpoint =
+            RpcUrl::parse(server.base_url().as_str()).context("parsing blob server URL")?;
+        self.blob_server = Some(server);
+        Ok(endpoint)
+    }
+
+    /// Return the blob server URL if it has been started.
+    pub fn blob_server_url(&self) -> Result<RpcUrl> {
+        let server = self.blob_server.as_ref().context("blob server not started")?;
+        RpcUrl::parse(server.base_url().as_str()).context("parsing blob server URL")
+    }
+
+    /// Return the blob server endpoint as a `url::Url`.
+    pub fn blob_server_endpoint(&self) -> Result<Url> {
+        let server = self.blob_server.as_ref().context("blob server not started")?;
+        Ok(server.endpoint().clone())
+    }
+
+    /// Explicit async teardown to revert the L1 snapshot.
+    pub async fn shutdown(self) -> Result<()> {
+        if let Some(server) = self.blob_server {
+            server.shutdown().await?;
+        }
+        revert_snapshot(&self.cleanup_provider, &self.snapshot_id).await
+    }
+}
+
+impl AsyncTestContext for ShastaEnv {
+    /// Setup the ShastaEnv before each test.
+    async fn setup() -> Self {
+        ShastaEnv::load_from_env().await.expect("failed to load ShastaEnv")
+    }
+
+    /// Teardown the ShastaEnv after each test.
+    async fn teardown(self) {
+        self.shutdown().await.expect("ShastaEnv teardown must succeed");
```

Comment:
```suggestion
        self.shutdown().await.unwrap_or_else(|e| panic!("ShastaEnv teardown failed: {e:?}"));
```

maybe something like that would be

Reply:
updated: https://github.com/taikoxyz/taiko-mono/pull/20866/commits/504fcf3222092116de8b6b75de29e82808766751